src/number.rs

```diff
@@ -292,90 +391,381 @@ impl<'de> Deserialize<'de> for Number {
+        self.deserialize_any(visitor)
+    }
+
+    #[cfg(feature = "arbitrary_precision")]
+    fn deserialize_str<V>(self, visitor: V) -> Result<V::Value, Error>
+    where
+        V: de::Visitor<'de>,
+    {
+        visitor.visit_str(&self.n)
+    }
+
+    #[cfg(not(feature = "arbitrary_precision"))]
+    fn deserialize_string<V>(self, visitor: V) -> Result<V::Value, Error>
+    where
+        V: de::Visitor<'de>,
+    {
+        self.deserialize_any(visitor)
+    }
+
+    #[cfg(feature = "arbitrary_precision")]
+    fn deserialize_string<V>(self, visitor: V) -> Result<V::Value, Error>
+    where
+        V: de::Visitor<'de>,
+    {
+        visitor.visit_str(&self.n)
+    }
+
+    forward_to_deserialize_any! {
+        bool char bytes byte_buf option unit unit_struct newtype_struct seq
+        tuple tuple_struct map struct enum identifier ignored_any
+    }
+}
+
+#[cfg(feature = "arbitrary_precision")]
+// Not public API. Should be pub(crate).
+#[doc(hidden)]
+pub struct NumberDeserializer<'a> {
+    pub visited: bool,
+    pub number: Cow<'a, str>,
+}
+
+#[cfg(feature = "arbitrary_precision")]
+impl<'de> MapAccess<'de> for NumberDeserializer<'de> {
+    type Error = Error;
+
+    fn next_key_seed<K>(&mut self, seed: K) -> Result<Option<K::Value>, Error>
+    where
+        K: de::DeserializeSeed<'de>,
+    {
+        if self.visited {
+            return Ok(None)
+        }
+        self.visited = true;
+        seed.deserialize(NumberFieldDeserializer).map(Some)
+    }
+
+    fn next_value_seed<V>(&mut self, seed: V) -> Result<V::Value, Error>
+    where
+        V: de::DeserializeSeed<'de>,
+    {
+        seed.deserialize(self.number.to_owned().into_deserializer())
+    }
+}
+
+#[cfg(feature = "arbitrary_precision")]
+struct NumberFieldDeserializer;
+
+#[cfg(feature = "arbitrary_precision")]
+impl<'de> Deserializer<'de> for NumberFieldDeserializer {
+    type Error = Error;
+
+    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Error>
+    where
+        V: de::Visitor<'de>,
+    {
+        visitor.visit_borrowed_str(SERDE_STRUCT_FIELD_NAME)
+    }
+
     forward_to_deserialize_any! {
-        bool i8 i16 i32 i64 u8 u16 u32 u64 f32 f64 char str string bytes
-        byte_buf option unit unit_struct newtype_struct seq tuple
-        tuple_struct map struct enum identifier ignored_any
+        bool u8 u16 u32 u64 i8 i16 i32 i64 f32 f64 char str string seq
+        bytes byte_buf map struct option unit newtype_struct
+        ignored_any unit_struct tuple_struct tuple enum identifier
     }
 }
 
-macro_rules! from_signed {
-    ($($signed_ty:ident)*) => {
-        $(
-            impl From<$signed_ty> for Number {
-                #[inline]
-                fn from(i: $signed_ty) -> Self {
-                    if i < 0 {
-                        Number { n: N::NegInt(i as i64) }
-                    } else {
-                        Number { n: N::PosInt(i as u64) }
-                    }
+impl FromPrimitive for Number {
```

Comment:
The num-traits crate is not stable yet so we cannot have these trait impls in the serde_json public API. Please find a different way to implement this.

Reply:
Could we just hide the impl's for them?

Reply:
I would prefer to not have these impls in the public API even if we hide them. Is it possible to implement a different way?

Reply:
Yes, although `FromPrimitive`/`ToPrimitive` does the heavy lifting for us, which is quite nice. I also don't see what's "unstable" about them... is the fact the version is 0.2 really that bad? Serde was 0.x not so long ago, after all.

Reply:
I guess I don't understand what heavy lifting you are referring to. Just looking at the [current implementation](https://github.com/serde-rs/json/blob/v1.0.10/src/number.rs#L350) of the `From` impls it looks like making them work in the arbitrary_precision case should be not much more than 1 line of code: `Number { n: i.to_string() }`.

Reply:
This has not been addressed yet.

Reply:
See my comment in the main thread.